### PR TITLE
YJ-002: Create sales employee api endpoints

### PR DESCRIPTION
**Pull request by Aaron-Millen**

### CODE REVIEW

File: src/main/java/org/example/controllers/SalesEmployeeController.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                    .entity(e.getMessage()).build();
+        }
+
+
+    }
+
+    @DELETE
+    @Path("/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response salesEmployeeService(final @PathParam("id") int id) {
```
Comment (lukemKainos) [inline src/main/java/org/example/controllers/SalesEmployeeController.java]: what is this name

File: src/main/java/org/example/daos/SalesEmployeeDao.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class SalesEmployeeDao {
+    final int parameterOne = 1;
+    final int parameterTwo = 2;
+    final int parameterThree = 3;
+    final int parameterFour = 4;
+    final int parameterFive = 5;
+    final int parameterSix = 6;
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/SalesEmployeeDao.java]: dont like just saying paramn to get out of magic numbers, variables should be descriptive

Comment (lukemKainos) [reply src/main/java/org/example/daos/SalesEmployeeDao.java]: make them static, make them capital

File: src/main/java/org/example/daos/SalesEmployeeDao.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                        resultSet.getDouble("commissionRate")
+                );
+
+            }
+        }
+        return null;
+    }
+    public int createSalesEmployee(final SalesEmployeeRequest
+                                           salesEmployee) throws SQLException {
+        Connection c = DatabaseConnector.getConnection();
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/SalesEmployeeDao.java]: try with

File: src/main/java/org/example/daos/SalesEmployeeDao.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                return employeeId;
+            }
+        }
+
+        return -1;
+    }
+    public void updateSalesEmployee(final int id,
+                                    final SalesEmployeeRequest salesEmployee)
+            throws SQLException {
+        Connection c = DatabaseConnector.getConnection();
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/SalesEmployeeDao.java]: try with pls

File: src/main/java/org/example/daos/SalesEmployeeDao.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+            while (resultSet.next()) {
+                SalesEmployee salesEmployee = new SalesEmployee(
+                        resultSet.getInt("Employee.employeeId"),
+                        resultSet.getString("firstName"),
+                        resultSet.getString("lastName"),
+                        resultSet.getDouble("salary"),
+                        resultSet.getString("bankAccountNumber"),
+                        resultSet.getString("nationalInsuranceNumber"),
+                        resultSet.getDouble("commissionRate")
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/SalesEmployeeDao.java]: this is replicated a lot, should be abstracted

File: src/main/java/org/example/daos/SalesEmployeeDao.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.List;
+
+public class SalesEmployeeDao {
+    final int parameterOne = 1;
+    final int parameterTwo = 2;
+    final int parameterThree = 3;
+    final int parameterFour = 4;
+    final int parameterFive = 5;
+    final int parameterSix = 6;
+    public List<SalesEmployee> getAllSalesEmployees() throws SQLException {
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/SalesEmployeeDao.java]: bin the All

File: src/main/java/org/example/exceptions/Entity.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package org.example.exceptions;
+
+public enum Entity {
+    SALESEMPLOYEE("SalesEmployee");
```
Comment (lukemKainos) [inline src/main/java/org/example/exceptions/Entity.java]: _ seperator pls

File: src/main/java/org/example/mappers/SalesEmployeeMapper.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                .map(salesEmployee -> new SalesEmployeeResponse(
+                        salesEmployee.getEmployeeId(),
+                        salesEmployee.getFirstName(),
+                        salesEmployee.getLastName(),
+                        salesEmployee.getSalary(),
+                        salesEmployee.getBankAccountNumber(),
+                        salesEmployee.getNationalInsuranceNumber(),
+                        salesEmployee.getCommissionRate()))
+                .collect(Collectors.toList());
+    }
```
Comment (lukemKainos) [inline src/main/java/org/example/mappers/SalesEmployeeMapper.java]: `public static List<SalesEmployeeResponse> toResponseList(List<SalesEmployee> salesEmployees) {
    return salesEmployees.stream()
            .map(SalesEmployeeResponse::new)
            .collect(Collectors.toList());
}`

File: src/main/java/org/example/models/EmployeeRequest.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class EmployeeRequest {
+    private String firstName;
+    private String lastName;
+    private double salary;
+    String bankAccountNumber;
+    String nationalInsuranceNumber;
```
Comment (lukemKainos) [inline src/main/java/org/example/models/EmployeeRequest.java]: private pls

File: src/main/java/org/example/models/EmployeeRequest.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
+package org.example.models;
```
Comment (lukemKainos) [inline src/main/java/org/example/models/EmployeeRequest.java:1]: add default constructor pls

File: src/main/java/org/example/models/EmployeeResponse.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
+package org.example.models;
+
+public class EmployeeResponse {
+    private int employeeId;
+    private String firstName;
+    private String lastName;
+    private double salary;
+    String bankAccountNumber;
+    String nationalInsuranceNumber;
```
Comment (lukemKainos) [inline src/main/java/org/example/models/EmployeeResponse.java]: private pls

File: src/main/java/org/example/models/EmployeeResponse.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    public EmployeeResponse(final int employeeId, final String firstName,
```
Comment (lukemKainos) [inline src/main/java/org/example/models/EmployeeResponse.java:10]: employeeId goes nowhere?

File: src/main/java/org/example/services/SalesEmployeeService.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public void deleteSalesEmployee(
+            final int id) throws SQLException, DoesNotExistException {
+        SalesEmployee salesEmployeeToUpdate =
+                salesEmployeeDao.getSalesEmployeeById(id);
+        if (salesEmployeeToUpdate == null) {
+            throw new DoesNotExistException(Entity.SALESEMPLOYEE);
+        }
+        salesEmployeeDao.deleteProduct(id);
```
Comment (lukemKainos) [inline src/main/java/org/example/services/SalesEmployeeService.java]: delete product doesnt make sense

File: src/main/java/org/example/services/SalesEmployeeService.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                .mapSalesEmployeeListToSalesEmployeeResponseList(
+                salesEmployeeDao.getAllSalesEmployees());
+    }
+
+    public SalesEmployee getSalesEmployeeByID(final int id) throws SQLException,
+            DoesNotExistException {
+        SalesEmployee salesEmployee = salesEmployeeDao.getSalesEmployeeById(id);
+        if (salesEmployee == null) {
+            throw new DoesNotExistException(Entity.SALESEMPLOYEE);
+        }
```
Comment (lukemKainos) [inline src/main/java/org/example/services/SalesEmployeeService.java]: lets abstract this to a function

File: src/main/java/org/example/services/SalesEmployeeService.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    public void updateSalesEmployee(final int id,
+                 final SalesEmployeeRequest salesEmployeeRequest)
+            throws InvalidException, DoesNotExistException, SQLException {
+        salesEmployeeValidator.validateSalesEmployee(
+                salesEmployeeRequest);
+        SalesEmployee salesEmployeeToUpdate =
+                salesEmployeeDao.getSalesEmployeeById(id);
+        System.out.println(id);
+        System.out.println(salesEmployeeToUpdate);
```
Comment (lukemKainos) [inline src/main/java/org/example/services/SalesEmployeeService.java]: ahem

File: src/main/java/org/example/validators/SalesEmployeeValidator.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+public class SalesEmployeeValidator {
+
+    public void validateSalesEmployee(
+            final SalesEmployeeRequest salesEmployeeRequest)
+        throws InvalidException {
+        final int maximumCommissionRate = 1;
+        final int maximumNameLength = 50;
+        final int maximumBankAccountNumberLength = 34;
+        final int maximumNationalInsuranceNumberLength = 9;
+        final int maximumSalaryLength = 15;
```
Comment (lukemKainos) [inline src/main/java/org/example/validators/SalesEmployeeValidator.java]: make these static

File: src/main/java/org/example/validators/SalesEmployeeValidator.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                >= maximumCommissionRate) {
+            throw new InvalidException(
+                    Entity.SALESEMPLOYEE,
+                    "Commission must not be greater than one"
+            );
+        }
+        if (salesEmployeeRequest.getFirstName()
+                .length() > maximumNameLength) {
+            throw new InvalidException(
+                    Entity.SALESEMPLOYEE,
```
Comment (lukemKainos) [inline src/main/java/org/example/validators/SalesEmployeeValidator.java]: _ separate the enum

**Review by lukemKainos — CHANGES_REQUESTED** [review]

**Review by lukemKainos — APPROVED** [review]

**Review by MatthewJKainos — APPROVED** [review]

**Pull request merged**